internal/provider/harness_base_resource.go

```diff
@@ -133,20 +140,68 @@ func (r *HarnessResource) ModifyPlan(ctx context.Context, req resource.ModifyPla
 	}
 }
 
-func (r *HarnessResource) ShouldSkip(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) bool {
-	inv := InventoryDataSourceModel{}
-	if diags := req.Config.GetAttribute(ctx, path.Root("inventory"), &inv); diags.HasError() {
-		return false
+// Read implements resource.Resource. This is intentionally a no-op.
+func (r *BaseHarnessResource) Read(context.Context, resource.ReadRequest, *resource.ReadResponse) {}
+
+func (r *BaseHarnessResource) create(ctx context.Context, req resource.CreateRequest, harness itypes.Harness) error {
+	// We unmarshal this explicitly because of the insanity of the framework not
+	// supporting embedded structs
+	var data BaseHarnessResourceModel
+	req.Plan.GetAttribute(ctx, path.Root("inventory"), &data.Inventory)
+	req.Plan.GetAttribute(ctx, path.Root("name"), &data.Name)
+	req.Plan.GetAttribute(ctx, path.Root("id"), &data.Id)
+	req.Plan.GetAttribute(ctx, path.Root("timeouts"), &data.Timeouts)
+
+	return r.do(ctx, data, harness)
+}
+
+func (r *BaseHarnessResource) update(ctx context.Context, req resource.UpdateRequest, harness itypes.Harness) error {
+	// We unmarshal this explicitly because of the insanity of the framework not
+	// supporting embedded structs
+	var data BaseHarnessResourceModel
+	req.Plan.GetAttribute(ctx, path.Root("inventory"), &data.Inventory)
+	req.Plan.GetAttribute(ctx, path.Root("name"), &data.Name)
+	req.Plan.GetAttribute(ctx, path.Root("id"), &data.Id)
+	req.Plan.GetAttribute(ctx, path.Root("timeouts"), &data.Timeouts)
+
+	return r.do(ctx, data, harness)
+}
+
+func (r *BaseHarnessResource) do(ctx context.Context, data BaseHarnessResourceModel, harness itypes.Harness) error {
+	if r.skip(ctx, data.Inventory, data.Id.ValueString()) {
+		log.Info(ctx, "skipping harness", "id", data.Id.ValueString(), "reason", "feature labels do not match")
+		return nil
 	}
 
-	var id string
-	if diags := req.Plan.GetAttribute(ctx, path.Root("id"), &id); diags.HasError() {
-		return false
+	// NOTE: This is technically different for create/update, but we reuse the
+	// create timeouts everywhere
+	timeout, diags := data.Timeouts.Create(ctx, defaultHarnessCreateTimeout)
+	if diags.HasError() {
+		// Warn and default to the default timeout
+		log.Warn(ctx, "failed to parse harness create timeout", "error", diags)
+		timeout = defaultHarnessCreateTimeout
 	}
 
-	feats, err := r.store.Inventory(inv).GetFeatures(ctx, inventory.Harness(id))
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	r.store.AddHarness(data.Id.ValueString(), harness)
+
+	ctx, err := r.store.Logger(ctx, data.Inventory, "harness_id", data.Id.ValueString(), "harness_name", data.Name.ValueString())
+	if err != nil {
+		return fmt.Errorf("failed to initialize logger(s): %w", err)
+	}
+
+	if _, err := harness.Setup()(ctx); err != nil {
```

Comment:
this signature will be changed in the next PR, which will simplify the `Harness` interface